src/embed/connection.rs

```diff
@@ -0,0 +1,256 @@
+use super::dispatch::Request;
+use super::types::*;
+use crate::dag;
+use crate::db;
+use async_fn::AsyncFn2;
+use async_std::stream::StreamExt;
+use async_std::sync::{Receiver, RecvError, RwLock};
+use futures::stream::futures_unordered::FuturesUnordered;
+use log::warn;
+use nanoserde::{DeJson, SerJson};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU32, Ordering};
+
+lazy_static! {
+    static ref TRANSACTION_COUNTER: AtomicU32 = AtomicU32::new(1);
+}
+
+type TxnMap<'a> = RwLock<HashMap<u32, RwLock<db::Write<'a>>>>;
+
+fn deserialize<T: DeJson>(data: &str) -> Result<T, String> {
+    match DeJson::deserialize_json(data) {
+        Ok(v) => Ok(v),
+        Err(e) => Err(format!("InvalidJson({})", e)),
```

Comment:
So here's a counterargument: I'm routing has/get/put (and later del) though the `execute()` function to not repeat the JSON parsing, transaction lookup, and error formatting multiple times, and then since that's a generic function, you couldn't see the error type it was using anyways.

I used to have `execute()` take a third template parameter that was the forwarded function's error type, but only `get` had any errors it could return apart from this JSON parsing error; `has` and `put`'s error enums were otherwise empty (`put` used to have errors, but those are now part of `commit`).

I can maybe buy your case at the `db` level, but here in dispatch these `do_XXX` functions are bound to turn the error into a string to pass to JSON momentarily anyways, so having a bunch of error boilerplate and mapping code seemed overboard. I left the error enums for `OpenTransactionError` and `CommitError`, but no one is going to look at these.

I do agree that typed errors are nice if inspection is expected, but I believe that at this level all these causes are opaque to the `embed` module, unlike how, say, a `dag` error may be meaningful to the `db` module.

Reply:
> since that's a generic function, you couldn't see the error type it was using anyways.

I think it would be great if these JSON errors as an example were propagated up through the call signatures and you could see that they could happen statically.

> I used to have execute() take a third template parameter that was the forwarded function's error type

That would be super cool.

> only get had any errors it could return apart from this JSON parsing error

The others will grow errors over time. They only have no errors now due to the simplicity of the current impl of prolly.

> I can maybe buy your case at the db level, but here...

I think we should bias toward consistency here for a few reasons:

* It's one less thing to think about. Is this a case where we use statically typed errors or stringly typed errors? Blech. Just do it the same way everywhere, don't have to worry about it.
* It makes the code easier to read. You get used to a certain convention.
* It makes the code easier to refactor. If the callers of a function change, we don't have to change the error convention.

I agree the current pattern is a bit more laborious than ideal, perhaps there are ways to improve it without losing its benefits.

Reply:
Basically I would like to converge on an error convention system wide, and I really like the benefits of statically known errors.